src/nvim/api/vim.c

```diff
@@ -2851,3 +2852,103 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
 error:
   clear_provider(p);
 }
+
+///
+/// Set matches and start insert completion: see |complete()|.
+///
+/// @param startcol: Byte offset where the completed text starts.
+///                     Any text between startcol and the current
+///                     cursor col is replaced by the completion.
+/// @param matches: A list of string matches.
+/// @param opts: Dictionary. Possible keys:
+///                 filterfunc: function used for filtering the
+///                     matches. See |nvim_register_filterfunc|.
+//
+// todo: completefunc with refresh: always
+// todo: figure out what to do with ctrl_e and ctrl_l
+// todo: what to do if filterfunc != NULL AND completion entry
+// has icase/equal set.
+String nvim_complete(Integer startcol,
+                   Object matches,
+                   DictionaryOf(LuaRef) opts,
+                   Error *err)
+  FUNC_API_SINCE(7)
+{
+  typval_T tv;
+
+  if (matches.type != kObjectTypeArray) {
+    api_set_error(err, kErrorTypeValidation,
+              "wrong type");
+  }
+
+  if ((State & INSERT) == 0) {
+    api_set_error(err, kErrorTypeException,
+            "complete() can only be used in Insert mode");
+    goto error;
+  }
+
+  // Check for undo allowed here, because if something was already inserted
+  // the line was already saved for undo and this check isn't done.
+  if (!undo_allowed())
+      goto error;
+
+  if (startcol <= 0) {
+      goto error;
+  }
+
+  for (size_t i = 0; i < opts.size; i++) {
+    String k = opts.items[i].key;
+    Object *v = &opts.items[i].value;
+
+    if (strequal(k.data, "filterfunc")){
+      if (v->type != kObjectTypeLuaRef) {
+        api_set_error(err, kErrorTypeValidation,
+                "expected lua function");
+        goto error;
+      }
+      active_filterfunc = user_filterfunc = v->data.luaref;
+      v->data.luaref = LUA_NOREF;
+    } else {
+      api_set_error(err, kErrorTypeValidation,
+              "undexpected key");
+      goto error;
+    }
+  }
+
+  object_to_vim(matches, &tv, err);
+  if ERROR_SET(err){
+      goto error;
+  }
+
+  set_completion((colnr_T)startcol - 1, tv.vval.v_list);
+
+error:
+  tv_clear(&tv);
+
+  String ret = STRING_INIT;
+  return ret;
+}
+
+///
+/// Register a lua function for filtering completion matches.
+/// The function must take in two arguments: the prefix (the
+/// currently typed text), and the text for a completion entry, and
+/// returns a float. Higher return values imply a better match, and a
+/// return value of 0 removes the entry from being shown.
+/// To remove the currently registered function, call the function
+/// with a nil value.
+///
+/// @param func: function to register for filtering.
+///
+void nvim_register_filterfunc(LuaRef func, Error *err)
```

Comment:
I tried to take `func` to be an Object instead, but when I do that, I keep getting these errors `cannot convert given type` whenever I try to pass in a function.

Reply:
yes this should be a `LuaRef`. In this case we have to use `api_copy_luaref` for now.